src/Pelagos/Bundle/AppBundle/Command/DoiDatacitePreMigrationCommand.php

```diff
@@ -0,0 +1,195 @@
+                    echo 'Setting DOI status ' . $dataset->getId() . ' (' . $dataset->getUdi() . ") to public.\n";
+                } else {
+                    echo 'Already in the public state ' . $dataset->getId() . ' (' . $dataset->getUdi() . ")\n";
+                }
+            }
+        }
+    }
+
+    /**
+     * Validate publish for DOI for the dataset.
+     *
+     * @param Dataset $dataset   The dataset.
+     * @param string  $doiStatus The status of the DOI for the dataset.
+     *
+     * @return boolean
+     */
+    private function validatePublish(Dataset $dataset, $doiStatus)
+    {
+        if (in_array(
+            $dataset->getMetadataStatus(),
+            [DatasetSubmission::METADATA_STATUS_SUBMITTED, DatasetSubmission::METADATA_STATUS_BACK_TO_SUBMITTER]
+        )
+        ) {
+            // Need to publish all the datasets which are submitted and in reserved state
+            return true;
+        } elseif ($dataset->getMetadataStatus() === DatasetSubmission::METADATA_STATUS_ACCEPTED and
+            $dataset->getDatasetSubmission()->getRestrictions() === DatasetSubmission::RESTRICTION_RESTRICTED and
+            $doiStatus === DOI::STATUS_RESERVED) {
+            // Need to publish accepted datasets which are restricted and doi status is reserved.
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Deletes DOI property from Dataset and object for DIF approved datasets.
+     *
+     * @param EntityManager $entityManager An instance of EntityManager.
+     *
+     * @throws ORMException        Exception thrown when get DOI fails.
+     * @throws ReflectionException Exception thrown when Reflection class fails.
+     *
+     * @return void
+     */
+    private function deleteDoiDif(EntityManager $entityManager)
+    {
+        $datasets = $entityManager->getRepository(Dataset::class)->findBy(array(
+            'metadataStatus' => array(
+                DatasetSubmission::METADATA_STATUS_NONE
+            )
+        ));
+
+        foreach ($datasets as $dataset) {
+            if ($dataset->getIdentifiedStatus() === DIF::STATUS_APPROVED) {
+                // Using reflection class for attributes which do not have setters.
+                $doi = $dataset->getDoi();
+                try {
+                    $datasetReflection = new \ReflectionClass($dataset);
+                    $doiReflection = $datasetReflection->getProperty('doi');
+                    $doiReflection->setAccessible(true);
+                    $doiReflection->setValue($dataset, null);
+                } catch (\ReflectionException $exception) {
+                    throw new ReflectionException('Reflection class failed ' . $exception->getMessage());
+                }
+
+                try {
+                    $entityManager->persist($dataset);
+                    $entityManager->remove($doi);
+                    $entityManager->flush($dataset);
+                    $entityManager->flush(DOI::class);
```

Comment:
The flush method in EntityManager class takes in entity as an argument, but can be null. It looks like i need only one flush. 👍

Reply:
I looked it up, because I was always confused about it.
Indeed, flush takes an argument of entity. This is what is says about it:
>If an entity is explicitly passed to this method only this entity and the cascade-persist semantics + scheduled inserts/removals are synchronized.

And like you said, this is probably not needed here. Also ::class produces a class name (string) not object, so that was probably also not intended.